### ISSUE TITLE
loading sparse_categorical_crossentropy as custom loss function gives different results to loading normally

### ISSUE THREAD

**Issue by sbland** [keras-team/keras#11840]
I was going to setup my own loss function but in testing loading a custom function (using example here:https://github.com/keras-team/keras/blob/master/keras/losses.py) I found that the model only reached 0.1 accuracy.

```
import keras

mnist = keras.datasets.mnist

(x_train, y_train),(x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

from keras import backend as K
modelb = keras.models.Sequential([
  keras.layers.Flatten(),
  keras.layers.Dense(512, activation=keras.activations.relu),
  keras.layers.Dropout(0.2),
  keras.layers.Dense(10, activation=keras.activations.softmax)
])

modelb.compile(optimizer='adam',
              loss=keras.losses.sparse_categorical_crossentropy,
              metrics=['accuracy'])


modelb.fit(x_train, y_train, epochs=5)
modelb.evaluate(x_test, y_test)
```
```
[OUTPUT]:[0.06683433205941983, 0.9796]
```
```
def lossfunctopt(y_true, y_pred):
    return K.sparse_categorical_crossentropy(y_true, y_pred)

modelb.compile(optimizer='adam',
              loss=lossfunctopt,
              metrics=['accuracy'])


modelb.fit(x_train, y_train, epochs=5)
modelb.evaluate(x_test, y_test)
```
```
[OUTPUT]: [0.07013783397653606, 0.0982]

```

Is there something I'm missing here?

**Comment by gabrieldemarmiesse** [comment]
The generation of the inital weights is not deterministic. Did you take that into account?

**Comment by sbland** [comment]
> The generation of the inital weights is not deterministic. Did you take that into account?

Yes I ran a number of times each with a similar result.

**Comment by gabrieldemarmiesse** [comment]
Because, I mean, the numbers look pretty close. Can you transfer the weights from one model to the other and run the evaluate function?

**Comment by sbland** [comment]
> Because, I mean, the numbers look pretty close. Can you transfer the weights from one model to the other and run the evaluate function?

The loss is close but the accuracy is out by a factor of 10
[OUTPUT]:[0.06683433205941983, 0.9796]
[OUTPUT]:[0.07013783397653606, 0.0982]

**Comment by sbland** [comment]
> Oups, my bad. Indeed you're right. Can you look at the predictions in both cases and make sure the weights are the same in both models? To see if the problem is in the network or in the loss function or in the computing of the accuracy. That would be very helpful to debug. You can use `save_weights` and `load_weights`.

Sorry I copied the wrong bit initially. Loss function on first model is ``keras.losses.sparse_categorical_crossentropy``

Still same issue though. I have tried copying the weights and still out by a factor of 10.
Thanks for looking into this!
